### ISSUE TITLE
music_native: MIDI voices switching to ACGPIANO randomly

### ISSUE THREAD

**Issue by ohrrpgce-bugbot** [ohrrpgce/ohrrpgce#244]
[bz#244]

While playing my game, a midi voice played the wrong instrument.

It was only one of the four voices playing at the time.
I've also noticed that it was using the ACGPIANO voice.

I'm wondering if this is related to the other midi voice bug because:
Whenever I experience the other midi voice bug while importing music, I'm not hearing the song use the voices of the previously selected song (as reported).
What I hear is the song being played using the ACGPIANO voices.
I wonder if perhaps custum.exe is defaulting to the first listed instrument (ACGPIANO) for some reason.

I don't know how to reproduce this effect.
I've only once heard this happen in game, and it occasionally occurs while importing music in custum.

From: **Bagne &lt;Robin.Geodesic&gt;**
Reported version: 20060803 Hasta-la-qb+
Operating system:   Windows XP

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @arperry

Drum tracks converted from BAM often wind up as piano in MIDI. Are you importing MIDI files or are you using BAM files? (Strictly speaking, ACGPIANO is a BAM voice.)

Does it sometimes play the right instrument or is it always the same voice that gets switched over? Does it switch during the song or does it use the piano voice from the start of the song?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

I can confirm this bug. Occasionally, the MIDI player will switch a channel to #⁠0 (Acoustic Grand Piano) (yes, AC G Piano). I'm not sure why, and though previous investigations of this turned up nothing, I suspect there's a minor bug in how non-note events are processed. Not an off-by-one error, as that would bork the whole song, but... dunno.

It seems to occur randomly, too. Like, in Game, ocassionally, it'll just play a song with Piano, until you leave and come back. Then, it'll be normal again.

If we had a song that consistantly reproduced this bug, then I'd likely be able to fix it. But, we don't. It's random, and can strike any particular song.

I think that I need to put aside a big chunk of time wherein I will tear the engine apart, and figure out why it's doing this. Like, use a few songs which I know don't use piano, and then trace all the instrument changes, to see if it switches to piano.

But, not now. Later.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: **Bagne &lt;Robin.Geodesic&gt;**


> Drum tracks converted from BAM often wind up as piano in MIDI. Are you
> importing MIDI files or are you using BAM files? (Strictly speaking, ACGPIANO
> is a BAM voice.)
I've been importing BAM files.
It has occured two or three times now on the same song (my forest music for a particular map).


> Does it sometimes play the right instrument or is it always the same voice that
> gets switched over?
It only happens occasionally (maybe 2-3 times), and so far it's only voice #⁠1 that has been affected. 

> Does it switch during the song or does it use the piano
> voice from the start of the song?
When it happens, the voice is wrong from the beginning of the song.  It doesn't fix itself until you enter and exit a battle (i.e. the song is restarted).  If you change maps that share the same music, it doesn't change. I'm not sure yet what happens if you enter a door leading to another map (with different music), then go back.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: **Bagne &lt;Robin.Geodesic&gt;**

I reported this bug a long time ago, and I haven't had any problems with in recent months.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen


> I reported this bug a long time ago, and I haven't had any problems with in
> recent months.

I assume that when you reported it, the default music backend was music_native (and that's why the bug is categorised as music_native). Because of this bug and other MIDI bugs, the default music backend was changed to music_sdl, and there is no reason to believe that this was fixed.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

Created attachment 190
"Battle of Fanfare", taken from Chronoboy Adventures

I've found that this bug very easy to reproduce using a number of different MIDIs. One of these is attached. Just change song to a different MIDI within a fraction of a second from the start of the song, and it happens about 70% of the time for me.

> Attached file: [Battle of Fanfare.mid](https://rpg.hamsterrepublic.com/bugzilla/old-attachments/Battle of Fanfare.mid) (audio/midi, 748 bytes)
> Description:   "Battle of Fanfare", taken from Chronoboy Adventures
